compiler/map.go

```diff
@@ -263,3 +267,73 @@ func hashmapIsBinaryKey(keyType types.Type) bool {
 		return false
 	}
 }
+
+func (b *builder) zeroUndefBytes(typ types.Type, ptr llvm.Value) error {
+	// We know that hashmapIsBinaryKey is true, so we only have to handle those types that can show up there.
+	// To zero all undefined bytes, we iterate over all the fields in the type.  For each element, compute the
+	// offset of that element.  If it's Basic type, there are no internal padding bytes.  For compound types, we recurse to ensure
+	// we handle nested types.  Next, we determine if there are any padding bytes before the next
+	// element and zero those as well.
+
+	switch typ := typ.Underlying().(type) {
+	case *types.Basic:
+		// no padding bytes
+		return nil
+	case *types.Pointer:
+		// mo padding bytes
+		return nil
+	case *types.Named:
+		// zero underlying type
+		return b.zeroUndefBytes(typ.Underlying(), ptr)
+	case *types.Array:
+		llvmArrayType := b.getLLVMType(typ)
+		llvmElemType := b.getLLVMType(typ.Elem())
+		base := ptr
+
+		for i := int64(0); i < typ.Len(); i++ {
+			// for each element, first clear any undef bytes in the element itself
+			idx := llvm.ConstInt(b.uintptrType, uint64(i), false)
+			ptr := b.CreateGEP(llvmArrayType, base, []llvm.Value{idx}, "")
+
+			// recursively zero any padding bytes in this element
+			b.zeroUndefBytes(typ.Elem(), ptr)
+
+			// check for padding between elements
+			// TODO(dgryski): typeSizeEqualStoreSize ?
+			if allocSize, storeSize := b.targetData.TypeAllocSize(llvmElemType), b.targetData.TypeStoreSize(llvmElemType); allocSize != storeSize {
+				n := llvm.ConstInt(b.uintptrType, allocSize-storeSize, false)
+				llvmSize := llvm.ConstInt(b.uintptrType, storeSize, false)
+				paddingStart := b.CreateGEP(b.ctx.Int8Type(), ptr, []llvm.Value{llvmSize}, "")
+				b.createRuntimeCall("memzero", []llvm.Value{paddingStart, n}, "")
+			}
+		}
+
+	case *types.Struct:
+		llvmStructType := b.getLLVMType(typ)
+		base := ptr
+
+		for i := 0; i < typ.NumFields(); i++ {
+			offset := b.targetData.ElementOffset(llvmStructType, int(i))
+			llvmOffset := llvm.ConstInt(b.uintptrType, offset, false)
+			ptr := b.CreateGEP(b.ctx.Int8Type(), base, []llvm.Value{llvmOffset}, "")
```

Comment:
Suggestion (I think this is less confusing):

```suggestion
			elemPtr := b.CreateGEP(b.ctx.Int8Type(), ptr, []llvm.Value{llvmOffset}, "")
```